### DISCUSSION TITLE
Adding attributes to root span

### DISCUSSION THREAD

**Discussion by piotrplazienski** [open-telemetry/opentelemetry-java-instrumentation#4767] [Q&A] [answered]
Hello,
We have migrated recently with our instrumentation from Elastic APM SDK to open telemetry (using java agent), with traces sent to Elastic APM.
It is working correctly, but we have problem with how Elastic APM interprets attributes sent by open telemetry.
Elastic APM treats root span (`processor.event: transaction`) differently than other spans (`processor.event: span`). Among others it allows to search for transactions by attributes (or labels in APM).
For basic server scenario, otel agent creates two spans: one for HTTP request (this one will be "transaction" in APM), and one child for Controller method.
So when I call `Span.current()` and add attributes to it, they will be added to child span (and later become `labels.xxx` in APM). From open telemetry perspective everything is ok, as attributes are propagated properly, but we cannot search for this in APM by labels/attributes.

Solution would be to add attributes to root span (one representing HTTP request, which later becomes `transaction`), but I failed to find any API that allows to obtain reference to it.
So question is: how to add attributes to root span in java agent (spring boot specifically, but maybe there's generic answer), or maybe there's some other way to propagate attributes so they are visible in top level APM?

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @piotrplazienski ,
The `opentelemetry-instrumentation-api` artifact contains a [`ServerSpan`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/ServerSpan.java) class that makes it possible to retrieve the _local_ root span ("local" means "the first span in this trace created by the current process" here). Does that fit your scenario?
Keep in mind that this module is still alpha and things may change - although I believe we're definitely not going to remove the `ServerSpan` class, it's very useful concept in general.

**Reply by piotrplazienski** [reply]
Hi @mateuszrzeszutek, this is _exactly_ what I needed, _local_ root span. I have verified that code `ServerSpan.fromContextOrNull(Context.current())?.setAttribute("someAttr", "someValue")` sets attributes to top level span, and Elastic APM picks this up as transaction labels.

I second `ServerSpan` as useful concept - from what I know not only Elastic APM treats root span in special way.
Thank you for fast response.
